h/schemas/forms/admin/organization.py

```diff
@@ -2,21 +2,38 @@
 
 import colander
 from deform.widget import TextAreaWidget, TextInputWidget
+from xml.etree import ElementTree
 
 import h.i18n
 from h import validators
-from h.models.organization import Organization, ORGANIZATION_NAME_MIN_CHARS, ORGANIZATION_NAME_MAX_CHARS
 from h.schemas.base import CSRFSchema
 
 _ = h.i18n.TranslationString
 
 
+ORGANIZATION_NAME_MIN_CHARS = 1
+ORGANIZATION_NAME_MAX_CHARS = 25
+ORGANIZATION_LOGO_MAX_CHARS = 100000
```

Comment:
Putting the field length constraints here introduces an inconsistency compared to `h.models.user` and `h.models.group`. The pattern we have ended up with for those other models is:

- The validation is actually implemented in three places: Once in the schema for colander forms, once in the JSON schema for the API and once in the domain model itself. The duplication is perhaps not ideal but not causing us that much pain either since it is quite simple.
- The values (min/max field lengths etc.) are at least defined in one central place in the `h.models.{group, user}` package.

Unless we think the way it works for `Group` and `User` is broken and we intend to change it, I would be inclined to follow the same pattern and leave the constants in `h.models.organization`. In this case we only have a form rather than an API _currently_ so we only need to use the values in this module. Thoughts @seanh ?

Reply:
I think you need to ask whether the min and max chars is a simple and universal constraint or not.

I think the general principle is that if a piece of validation is both really simple and universal (definitely going to want to apply _whenever_ we add anything to this DB table) then it belongs in the models. Examples: foreign key constraints, not null constraints, etc. Otherwise if it's more complicated or non-universal (e.g. different validation for API vs views, or for one type of user vs another, ...) then it belongs in `h.schemas` (which is planned to be renamed to `h.validation`) because (for several reasons) we try to keep our model classes really thin and simple.

Most of the time these really simple validations can be implemented as actual database constraints, but sometimes they can't be or its for some reason undesirable to implement that at Postgres-level, in those cases we can use `@sa.orm.validates()` methods on model classes. Uses of `@sa.orm.validates()` should be rare in our code though.

For string lengths we use `@sa.orm.validates()` when we put them in the models. We actually went through all the DB tables and replaced DB-level max lengths with `@sa.orm.validates()`. See [No length limits on database columns](http://h.readthedocs.io/en/latest/developing/making-changes-to-model-code/#no-length-limits-on-database-columns).

... So that's why things like `USERNAME_MAX_LENGTH` are in the models. And then of course the schemas need to give a friendly error message when something is too long, and such human-readable error message belong in the schemas or views not in the models, so that's why schemas need to import `USERNAME_MAX_LENGTH`.

Reply:
Sounds like they need to go back in the model then-will do!

Reply:
It does seem weird to me that the min and max chars check happens both in the schema and the model and it's not re-used. Ideally it should happen once in the most externally facing place to fail quickly.